src/main/java/com/iota/iri/utils/dag/DAGUtils.java

```diff
@@ -0,0 +1,194 @@
+package com.iota.iri.utils.dag;
+
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.storage.Tangle;
+
+import java.util.*;
+
+/**
+ * This class offers generic functions for recurring tasks that are related to the tangle and that otherwise would have
+ * to be implemented over and over again in different parts of the code.
+ */
+public class DAGUtils {
+    /**
+     * Holds references to the singleton DAGUtils instances.
+     */
+    protected static HashMap<Tangle, DAGUtils> instances = new HashMap<>();
+
+    /**
+     * Holds a reference to the tangle instance which acts as an interface to the used database.
+     */
+    protected Tangle tangle;
+
+    /**
+     * This method allows us to retrieve the DAGUtils instance that corresponds to the given parameters.
+     *
+     * To save memory and to be able to use things like caches effectively, we do not allow to create multiple instances
+     * of this class for the same set of parameters, but create a singleton instance of it.
+     *
+     * @param tangle database interface used to retrieve data
+     * @return DAGUtils instance that allows to use the implemented methods.
+     */
+    public static DAGUtils get(Tangle tangle) {
+        DAGUtils instance = instances.get(tangle);
+        if(instance == null) {
+            instance = new DAGUtils(tangle);
+
+            instances.put(tangle, instance);
+        }
+
+        return instance;
+    }
+
+    /**
+     * Constructor of the class which allows to provide the required dependencies for the implemented methods.
+     *
+     * Since the constructor is protected we are not able to manually create instances of this class and have to
+     * retrieve them through their singleton accessor {@link #get}.
+     *
+     * @param tangle database interface used to retrieve data
+     */
+    protected DAGUtils(Tangle tangle) {
```

Comment:
private

Reply:
protected to be able to still create manual utils objects for testing (test are in the same packet).